src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -1275,6 +1275,185 @@ int MacroAssembler::pop_fp(unsigned int bitset, Register stack) {
   return count;
 }
 
+static const int64_t right_32_bits = right_n_bits(32);
+static const int64_t right_8_bits = right_n_bits(8);
+
+/**
+ * Emits code to update CRC-32 with a byte value according to constants in table
+ *
+ * @param [in,out]crc   Register containing the crc.
+ * @param [in]val       Register containing the byte to fold into the CRC.
+ * @param [in]table     Register containing the table of crc constants.
+ *
+ * uint32_t crc;
+ * val = crc_table[(val ^ crc) & 0xFF];
+ * crc = val ^ (crc >> 8);
+ *
+ */
+void MacroAssembler::update_byte_crc32(Register crc, Register val, Register table) {
+  assert_different_registers(crc, val, table);
+
+  xorr(val, val, crc);
+  andi(val, val, right_8_bits);
+  shadd(val, val, table, val, 2);
+  lwu(val, Address(val));
+  srli(crc, crc, 8);
+  xorr(crc, val, crc);
+}
+
+/**
+ * Emits code to update CRC-32 with a 32-bit value according to tables 0 to 3
+ *
+ * @param [in,out]crc   Register containing the crc.
+ * @param [in]v         Register containing the 32-bit to fold into the CRC.
+ * @param [in]table0    Register containing table 0 of crc constants.
+ * @param [in]table1    Register containing table 1 of crc constants.
+ * @param [in]table2    Register containing table 2 of crc constants.
+ * @param [in]table3    Register containing table 3 of crc constants.
+ *
+ * uint32_t crc;
+ *   v = crc ^ v
+ *   crc = table3[v&0xff]^table2[(v>>8)&0xff]^table1[(v>>16)&0xff]^table0[v>>24]
+ *
+ */
+void MacroAssembler::update_word_crc32(Register crc, Register v, Register tmp1, Register tmp2,
+        Register table0, Register table1, Register table2, Register table3, bool upper) {
+  assert_different_registers(crc, v, tmp1, tmp2, table0, table1, table2, table3);
+
+  if (upper)
+    srli(v, v, 32);
+  xorr(v, v, crc);
+
+  andi(tmp1, v, right_8_bits);
+  shadd(tmp1, tmp1, table3, tmp2, 2);
+  lwu(crc, Address(tmp1));
+
+  // In order to access table elements according to initial algorithm
+  // the following actions should be performed (with no Zba enabled):
+  //  tmp1 = v >> 8
+  //  tmp1 = tmp1 & right_8_bits
+  //  tmp1 = tmp1 << 2
+  //  tmp1 += table2
+  // Which is the same as:
+  //  tmp1 = v >> 6
+  //  tmp1 = tmp1 & (right_8_bits << 2)
+  //  tmp1 += table2
+
+  srli(tmp1, v, 6);
+  andi(tmp1, tmp1, (right_8_bits << 2));
+  add(tmp1, tmp1, table2);
+  lwu(tmp2, Address(tmp1));
+
+  srli(tmp1, v, 14);
+  xorr(crc, crc, tmp2);
+
+  andi(tmp1, tmp1, (right_8_bits << 2));
+  add(tmp1, tmp1, table1);
+  lwu(tmp2, Address(tmp1));
+
+  if (upper) {
+    tmp1 = v;
+    srli(tmp1, v, 24);
+  }
+  else
+    srliw(tmp1, v, 24);
+  xorr(crc, crc, tmp2);
+
+  // no need to clear bits other than lowest two
+  shadd(tmp1, tmp1, table0, tmp1, 2);
+  lwu(tmp2, Address(tmp1));
+  xorr(crc, crc, tmp2);
```

Comment:
I witnessed slightly better JMH numbers on Lichee-PI-4A with the following sequence:
```
  if (upper)
    srli(v, v, 32);
  xorr(v, v, crc);

  andi(tmp1, v, right_8_bits);
  shadd(tmp1, tmp1, table3, tmp2, 2);
  lwu(crc, Address(tmp1));

  srli(tmp1, v, 6);
  andi(tmp1, tmp1, (right_8_bits << 2));
  add(tmp1, tmp1, table2);
  lwu(tmp2, Address(tmp1));

  srli(tmp1, v, 14);
  andi(tmp1, tmp1, (right_8_bits << 2));
  add(tmp1, tmp1, table1);
  xorr(crc, crc, tmp2);

  lwu(tmp2, Address(tmp1));
  srliw(tmp1, v, 24);
  shadd(tmp1, tmp1, table0, tmp1, 2);
  xorr(crc, crc, tmp2);

  lwu(tmp2, Address(tmp1));
  xorr(crc, crc, tmp2);
```

Reply:
Thanks for pointing it out! Also some additional accel can be achieved by using `srli` for '`upper`' cases: 

On Lichee-Pi:
`srliw` only
| Benchmark                      |  (count) |  Mode | Cnt   |  Score  |   Error |  Units |
| --------------------------------- | ------ | ----------- | --------- | -------- | --------- | ------ |
| CRC32.TestCRC32.testCRC32Update  |     64 | thrpt  | 12 | 6512.348 | 146.138 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |    128 | thrpt |  12 | 3408.306 | 279.986 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |    256 | thrpt |  12 | 1971.538 | 100.804 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |    512 | thrpt |  12 | 1040.091 |   3.426 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |   2048 | thrpt |  12 |  272.233 |   3.844 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |  16384 | thrpt |  12 |   33.781 |   1.961 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |  65536 | thrpt  | 12  |   8.399 |   0.042 | ops/ms |

`srli` and `srliw`
| Benchmark   |                     (count) |  Mode | Cnt |    Score   |  Error |  Units |
| ----------------- | ---------------------------- | ------ | ----- | ------------ | -------- | -------- |
| CRC32.TestCRC32.testCRC32Update  |     64 | thrpt  | 12 | 6561.674 | 104.461 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |    128 | thrpt |  12 | 3586.810 | 109.934 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |    256 | thrpt  | 12 | 2024.515 |  16.118 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |    512  | thrpt  | 12 | 1047.475 |  39.745 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |   2048 | thrpt |  12  | 274.006 |   0.809 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |  16384 | thrpt  | 12  |  34.746 |   0.203 | ops/ms |
| CRC32.TestCRC32.testCRC32Update  |  65536 | thrpt  | 12   |  8.405 |  0.064 | ops/ms |